zcash_client_backend/src/data_api/wallet/input_selection.rs

```diff
@@ -394,20 +394,25 @@ where
         // of funds selected is strictly increasing. The loop will either return a successful
         // result or the wallet will eventually run out of funds to select.
         loop {
-            #[cfg(feature = "orchard")]
-            let (sapling_input_total, orchard_input_total) = (
-                shielded_inputs.sapling_value()?,
-                shielded_inputs.orchard_value()?,
-            );
-
             #[cfg(not(feature = "orchard"))]
-            let orchard_input_total = NonNegativeAmount::ZERO;
-
-            let use_sapling =
-                !(sapling_outputs.is_empty() && orchard_input_total >= amount_required);
+            let use_sapling = true;
             #[cfg(feature = "orchard")]
-            let use_orchard =
-                !(orchard_outputs.is_empty() && sapling_input_total >= amount_required);
+            let (use_sapling, use_orchard) = {
+                let (sapling_input_total, orchard_input_total) = (
+                    shielded_inputs.sapling_value()?,
+                    shielded_inputs.orchard_value()?,
+                );
+
+                // Use Sapling inputs if there are no Orchard outputs or there are not sufficient
+                // Orchard outputs to cover the amount required.
+                let use_sapling =
+                    orchard_outputs.is_empty() || amount_required > orchard_input_total;
+                // Use Orchard inputs if there are insufficient Sapling funds to cover the amount
+                // reqiuired.
+                let use_orchard = !use_sapling || amount_required > sapling_input_total;
```

Comment:
I know this is out-of-scope for this release, but just to document it as an issue: this code might set either `use_sapling` or `use_orchard` to `false` because it decides that the amount required is satisfiable from the other pool, but actually it is not because of fees (even though the amount could potentially have been satisfied from both pills). The correct way to handle this will be to subtract the marginal fee from each note value when calculating balance, as @nuttycom suggested some time ago.

Reply:
This isn't a current bug, because this whole thing is in a loop that continues selecting value until it either has fees covered or runs out of notes.